src/lib/dialog/dialog-container.ts

```diff
@@ -61,20 +63,27 @@ export class MdDialogContainer extends BasePortalHost implements OnDestroy {
 
     let attachResult = this._portalHost.attachComponentPortal(portal);
 
-    // If were to attempt to focus immediately, then the content of the dialog would not yet be
-    // ready in instances where change detection has to run first. To deal with this, we simply
-    // wait for the microtask queue to be empty.
-    this._ngZone.onMicrotaskEmpty.first().subscribe(() => {
-      this._elementFocusedBeforeDialogWasOpened = document.activeElement;
-      this._focusTrap.focusFirstTabbableElement();
-    });
+    this._focusFirstTabbableElement();
 
     return attachResult;
   }
 
-  /** @docs-private */
   attachTemplatePortal(portal: TemplatePortal): Map<string, any> {
-    throw Error('Not yet implemented');
+    if (this._portalHost.hasAttached()) {
+      throw new MdDialogContentAlreadyAttachedError();
+    }
+
+    let attachResult = this._portalHost.attachTemplatePortal(portal);
+
+    this._focusFirstTabbableElement();
+
+    return attachResult;
+  }
+
+  attachTemplateRef<T>(templateRef: TemplateRef<T>) {
```

Comment:
I'm not sure I understand this correctly, but to be able to create a portal, you need ViewContainerRef, which is not accessible inside `dialog.ts`. In my previous PR I had `MdDialog.openFromPortal()`, but I think it a little too tedious for user to have to create a portal rather than to be just able to pass in `TemplateRef`. But I don't really mind either.